File: markdowner-core/src/main/java/org/jesperancinha/parser/markdowner/helper/TagHelper.java
package org.jesperancinha.parser.markdowner.helper;

import java.nio.CharBuffer;
import java.util.Arrays;

public class TagHelper {

    /**
     * Strips the first # characters from a string and removes eventually leading spaces
     *
     * @param tag Tag {@link String} to be sanitized
     * @return The sanitized tag {@link String}
     */
    public static String sanitizeTag(final String tag) {
        final String preSanitizedTag = removeStartingHashTags(tag);
        CharBuffer charBuff = CharBuffer.allocate(preSanitizedTag.length());
        int finalSize = 0;
        for (int i = 0; i < preSanitizedTag.length(); i++) {
            char charCandidate = preSanitizedTag.charAt(i);
            if (charCandidate < 256) {
                charBuff = charBuff.append(charCandidate);
                finalSize++;
            }
        }
        return new String(Arrays.copyOfRange(charBuff.array(), 0, finalSize)).trim();
    }

    private static String removeStartingHashTags(String tag) {
        int i = 0;
        for (char c : tag.toCharArray()) {
            if (c != '#') {
                return tag.substring(i).trim();
            }
            i++;
        }
        return "";
    }

    /**
     * Counts the number of leading hash characters in a {@link String}
     *
     * @param tag Tag {@link String} to analyse
     * @return Number of hash (#) characters
     */
    public static int counHashTags(final String tag) {
        int count = 0;
        for (char c : tag.toCharArray()) {
            if (c == '#') {
                count++;
            } else {
                return count;
            }
        }
        return count;
    }


}

File: markdowner-core/src/main/kotlin/org/jesperancinha/parser/markdowner/helper/TagHelper.kt
package org.jesperancinha.parser.markdowner.helper

import java.nio.CharBuffer
import java.util.*

object TagHelper {
    /**
     * Strips the first # characters from a string and removes eventually leading spaces
     *
     * @param tag Tag [String] to be sanitized
     * @return The sanitized tag [String]
     */
    fun sanitizeTag(tag: String?): String {
        val preSanitizedTag = removeStartingHashTags(tag)
        var charBuff = CharBuffer.allocate(preSanitizedTag.length)
        var finalSize = 0
        for (i in 0 until preSanitizedTag.length) {
            val charCandidate = preSanitizedTag[i]
            if (charCandidate.code < 256) {
                charBuff = charBuff.append(charCandidate)
                finalSize++
            }
        }
        return String(Arrays.copyOfRange(charBuff.array(), 0, finalSize)).trim { it <= ' ' }
    }

    private fun removeStartingHashTags(tag: String?): String {
        var i = 0
        for (c in tag!!.toCharArray()) {
            if (c != '#') {
                return tag.substring(i).trim { it <= ' ' }
            }
            i++
        }
        return ""
    }

    /**
     * Counts the number of leading hash characters in a [String]
     *
     * @param tag Tag [String] to analyse
     * @return Number of hash (#) characters
     */
    fun counHashTags(tag: String?): Int {
        var count = 0
        for (c in tag!!.toCharArray()) {
            if (c == '#') {
                count++
            } else {
                return count
            }
        }
        return count
    }
}